Set Up Plugin Tests

qclin opened the pull request:
The idea is to pull the latest Open Project image from [Docker hub ](https://hub.docker.com/r/openproject/community/tags) and then override the environment setting to match the ones from the CI Dockerfile to run the tests 

Findings 

- in attempt to isolate testing for plugin we found out that the ng test option --include does not look up symlinked files, which would actually exclude our kittens specs, one work around is to add focus to the top of our spec files this is a karma setting 

TODO 
- run only the plugin test for both frontend and backend 
- fix 
```
An error occurred while loading ./spec_legacy.
Failure/Error: __send__(method, file)

LoadError:
  cannot load such file -- /app/spec_legacy
```
- run linting only on the plugin files

qclin committed:
setup docker file

pmeinhardt reviewed (commented):
🤔  Looks like a lot of copied code from: https://github.com/opf/openproject/tree/dev/docker/ci

🗑️ If you're starting from the [openproject/community](https://hub.docker.com/r/openproject/community) image anyways, you should be able to remove the majority of this.

qclin commented:
yes it is, the openproject/community image uses the production docker setting, so we compare the difference between the Prod and CI Dockerfile and added the settings we needed

qclin committed:
add two simple component tests

qclin committed:
try github workflow

tessi commented on docker/ci/entrypoint.sh:
```diff
@@ -0,0 +1,91 @@
+if [ "$1" == "run-units" ]; then
+	shift
+	execute "cd frontend && npm install && npm run test"
+	execute "time bundle exec rspec -I spec_legacy spec_legacy"
+  # TODO: replace test command
+	if ! execute "time bundle exec rake -T" ; then
```
```suggestion
	if ! execute "time bundle exec rake parallel:plugins:specs" ; then
```

tessi replied:
I take my previous suggestion back (it would test all tests in *all* plugins -- not just our plugin) 🙈 

Instead, the following should work
```suggestion
	if ! execute "time bundle exec rspec --exclude-pattern '/plugin/spec/features/**/*' /plugin/spec/**/*_spec.rb" ; then
```

tessi commented on docker/ci/entrypoint.sh:
```diff
@@ -0,0 +1,91 @@
+
+if [ "$1" == "run-units" ]; then
+	shift
+	execute "cd frontend && npm install && npm run test"
```
it seems we *are* executing frontend tests exactly here (`npm run test`)

qclin replied:
yes we are, and we also need a github action workflow for this no?

tessi commented on docker/ci/entrypoint.sh:
```diff
@@ -0,0 +1,91 @@
+	execute "cd frontend; npm install ; cd -"
+	execute "bundle exec rake assets:precompile"
+	execute "cp -rp config/frontend_assets.manifest.json public/assets/frontend_assets.manifest.json"
+  # TODO: replace test command
+	if ! execute "time bundle exec rake -T" ; then
```
```suggestion
	if ! execute "time bundle exec rake parallel:plugins:features" ; then
```

tessi replied:
same idea here:

```suggestion
	if ! execute "time bundle exec rspec /plugin/spec/features/**/*_spec.rb" ; then
```

tessi commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,15 @@
+
+      - name: Frontend Test
+        run: |
+          cd frontend && npm ci && npm test -- --no-watch --no-progress --browsers=ChromeHeadlessCI
```
this can't work, right? we are currently sitting in the frontend directory. there is no openproject core linked. just from the frontend, we can't run the tests (as they need to run coming from the core).

qclin replied:
ja this is not working at all, we'll need to do the same flow as the docker file, but i haven't figure out how to do that in the workflow way yet

tessi commented:
The good news, is that you started a CI job (it ran the wrong thing, but it did something! 🥳 )

apfelfabrik committed:
Merge pull request #1 from bitcrowd/fix-path-instructions

Update import path and instructions

qclin committed:
restore test commands

qclin committed:
clean up docker

bluemoonecho committed:
clean up changes

tessi committed:
Merge pull request #5 from bitcrowd/clean-ups

Clean up changes

bluemoonecho committed:
Add an example of context menu for work packages

To add a context menu entry for single work packages, the frontend part
of the plugin must register it's context menu as outlined on `main.ts`

The menu entry depends on the `link` property that is defined on the backend
on `engine.rb`

On the backend part we can skip creating a link property in specific
applications. Eg. User not being allowed to see the context menu.

It's currently not possible to add context menu entries for multiple
selection of work packages.

bluemoonecho committed:
translation update

qclin committed:
add docker compose file

qclin committed:
update test workflow

qclin committed:
run on ubuntu

qclin committed:
combine build and unit

qclin committed:
add include

qclin committed:
run feature tests

tessi committed:
Merge pull request #4 from bitcrowd/feat/add-action-to-context-menu

Add an example of a context menu for work packages

qclin committed:
undo spec

qclin committed:
add linting

qclin committed:
exit

qclin committed:
naming

Commit:
run only plugin frontend tests

Commit:
download spec folder from git

Commit:
add rspec to test.yml

Commit:
separate angular and ruby tests

Commit:
fix spec folder download

Commit:
separate rspec features and unit tests

Commit:
fix indentation

Commit:
add spec folder to features test

Commit:
add space

Commit:
add relative path for features test

Commit:
fix rspec patterns

apfelfabrik committed:
Update Dockerfile and Gemfile

We're building the openproject container from scratch

apfelfabrik committed:
Adapt scripts to the new container folder structure

apfelfabrik committed:
Attempt to resolve the remaining permission issues

apfelfabrik committed:
Also chown /usr/local/bundle to dev

apfelfabrik committed:
chown bundle cache in entrypoint, not Dockerfile

It seems we have to chown after bundle install, and install can only
happen after plugin volume is mounted in docker-compose

iampeternguyen commented on docker/ci/Dockerfile:
```diff
@@ -0,0 +1,66 @@
+WORKDIR /home/$USER/openproject
+
+ENTRYPOINT [ "/usr/sbin/entrypoint.sh" ]
+CMD ["setup-tests", "bash"]
```
I know this is from the openproject ci file but i don't understand this.

From the docs, the CMD has three forms

The CMD instruction has three forms:

    CMD ["executable","param1","param2"] (exec form, this is the preferred form)
    CMD ["param1","param2"] (as default parameters to ENTRYPOINT)
    CMD command param1 param2 (shell form)

So i guess we're using number 2. But why are we passing bash to the entry point file? What's the purpose?

apfelfabrik replied:
This is only for when we run the container w/o any arguments. When we run it like that, it'll run the `entrypoint.sh setup-tests bash`, executing `setup-tests` first and then executing `bash` with that last fallback case in `entrypoint.sh`.

As to why: This will give you an interactive shell for exploration rather than actually run any tests. That is considered more valuable when developing, i.e., when you use the `Dockerfile` directly. If you actually intend to run tests, you can invoke them straight away like we do in the workflow.

You might challenge this approach, but this is in line with how the `Dockerfile` in core works. And as an added hint that this is better than running tests directly you could argue that you probably don't want to run _all_ the tests here anyway, and picking any particular tests over others in the `Dockerfile` would encode an opinion about their relative importance, which we can avoid by just running bash instead.

apfelfabrik committed:
chown also before bundle install

apfelfabrik committed:
Don't wrap chowns in execute

apfelfabrik commented:
Things we could still do:
* Documentation of what we did
* Pull some of the tasks out into a reusable step within github workflows, optmize, this might be a bit explorative. @tessi has some ideas here for reusability.
* Figure out whether we need to compile the assets (angular/webpack) before running the feature tests. Currently, we're only installing npm dependencies before the frontend tests.
* Maybe have a look at whether we can make the feature tests green.

Commit:
add documentation

Commit:
fix same job name

Commit:
rename jobs

Commit:
use plugin args for all test jobs

Commit:
update docs

Commit:
use actions from test actions repo

Commit:
remove reusable workflows

Commit:
test using repeated node

Commit:
explicitly use plugin name

apfelfabrik commented on docker/ci/DOCKER.md:
```diff
@@ -0,0 +1,25 @@
+
+The Docker CI is used to run linting and tests designed for a CI pipeline. Github workflows have been added to .github/workflows that will automatically run linting and tests for only the plugin.
+
+These tests use OpenProject Core `v12.0.1` by default. To change this, go to `Dockerfile` and change `OPENPROJECT_BRANCH` to your desired version number.
```
We could say _"use a recent OpenProject Core production release by default"_, putting actual version numbers in the documentation will make it to rot away quicker, since the next person updating the `Dockerfile` will most certainly forget to update the Markdown as well.

apfelfabrik commented on README.md:
```diff
@@ -421,69 +418,66 @@ By using `render_on`, the given variables are made available as locals to the pa
 
-* :model_changeset_scan_commit_for_issue_ids_pre_issue_update (changeset, issue)
-* :copy_project_add_member (new_member, member)
+- :model_changeset_scan_commit_for_issue_ids_pre_issue_update (changeset, issue)
+- :copy_project_add_member (new_member, member)
```
Why the hate towards asterisks? 😆

tessi committed:
Set up plugin tests (#9)

* attempt at a custom composite action

* without local docker files

* with correct vars

* use test steps

* moahr steps

* use bitcrowd repo

* fix branch

* use v1 tag

Co-authored-by: Peter Nguyen <peter@bitcrowd.net>

tessi commented on README.md:
```diff
@@ -146,11 +144,17 @@ A special `.travis.yml` and `Gemfile.plugins` are included which allow you to ha
 the specs for your plugin along with the core specs. This is a good way to test if your plugin
 works properly in conjunction with the core.
 
+**Docker CI**
+
+A `docker/ci` folder exists to automate the testing of both angular unit and rspec unit/features. Additionally, a `lint.yml` and `test.yml` file exists in .github/workflows to use Github Actions.
+
+For more documentation, see `docker/ci/DOCKER.MD`.
+
```
```suggestion
```

tessi committed:
Update README.md
